src/main/scala/scynamo/ScynamoEncoder.scala

```diff
@@ -12,119 +12,150 @@ import shapeless.tag.@@
 import software.amazon.awssdk.services.dynamodb.model.AttributeValue
 
 import java.time.Instant
-import java.util.UUID
+import java.util.{Collections, UUID}
+import scala.collection.compat._
+import scala.collection.immutable.Seq
 import scala.concurrent.duration.{Duration, FiniteDuration}
-import scala.jdk.CollectionConverters._
 
 trait ScynamoEncoder[A] { self =>
   def encode(value: A): EitherNec[ScynamoEncodeError, AttributeValue]
-
-  def contramap[B](f: B => A): ScynamoEncoder[B] = value => self.encode(f(value))
+  def contramap[B](f: B => A): ScynamoEncoder[B] =
+    ScynamoEncoder.instance(value => self.encode(f(value)))
 }
 
 object ScynamoEncoder extends DefaultScynamoEncoderInstances {
   def apply[A](implicit instance: ScynamoEncoder[A]): ScynamoEncoder[A] = instance
+
+  // SAM syntax generates anonymous classes because of non-abstract methods like `contramap`.
+  private[scynamo] def instance[A](f: A => EitherNec[ScynamoEncodeError, AttributeValue]): ScynamoEncoder[A] = f(_)
+
+  // Ignore `nul` for efficiency and GSI support.
```

Comment:
could we expand `GSI` to `Global Secondary Index`?  Not sure if the comment here really helps, maybe adding a dedicated test with a link to the issue would be better to know what invariant we want to ensure?